Removed SDL_RendererFlags

slouken opened the pull request:
The vsync flag has been replaced with an SDL_bool in SDL_CreateRenderer() and SDL_PROP_RENDERER_VSYNC_NUMBER.  SDL_SetRenderVSync() now takes additional values besides 0 and 1.

The maximum texture size has been removed from SDL_RendererInfo, replaced with SDL_PROP_RENDERER_MAX_TEXTURE_SIZE_NUMBER.

slouken commented:
@icculus, can you give this a sanity check?

icculus commented:
Is there a scenario where we can set vsync at create time but can't toggle vsync later with SDL_SetRenderVSync? I'm wondering if we shouldn't just _remove_ the parameter and default to vsync, and those that care about it can manually turn it off after creating the renderer.

icculus commented:
SDL_CreateRenderer is one of those weird ones where the right misfortunes of compiler warnings will let this compile SDL2 code incorrectly, casting the second argument from an int to a pointer, so I really like the idea of removing a parameter so people definitely catch this, fwiw.

slouken commented:
> Is there a scenario where we can set vsync at create time but can't toggle vsync later with SDL_SetRenderVSync? I'm wondering if we shouldn't just _remove_ the parameter and default to vsync, and those that care about it can manually turn it off after creating the renderer.

No, it can be adjusted later in all renderers. We can remove the parameter, but if we do that, I think we should default vsync off, since people like to get the highest FPS possible.

slouken commented:
@icculus, I removed the flags parameter and adjusted the logic for setting vsync during renderer creation. Can you do a careful review of this commit?

sezero commented on docs/README-visualc.md:
```diff
@@ -88,7 +88,7 @@ Here's a sample SDL snippet to verify everything is setup in your IDE:
         SDL_Init(SDL_INIT_VIDEO);
         window = SDL_CreateWindow("Hello SDL", WIDTH, HEIGHT, 0);
-        renderer = SDL_CreateRenderer(window, NULL, SDL_RENDERER_PRESENTVSYNC);
+        renderer = SDL_CreateRenderer(window, NULL, SDL_TRUE);
```
This one is wrong after last parameter removal

slouken replied:
Fixed, thanks.

icculus commented on docs/README-winrt.md:
```diff
@@ -88,9 +88,9 @@ Here is a rough list of what works, and what doesn't:
-    WinRT (such as Windows 8.x), where turning off VSync appears to work.
+    forcing VSync back on.  As such, vsync will always get turned-on on Windows
+    Phone.  This limitation is not present in non-Phone WinRT (such as Windows 8.x),
+    where turning off VSync appears to work.
```
This isn't really a detail for this pull request, but: didn't we drop Windows Phone support?

slouken replied:
I think so, but I didn't want to assume here.

icculus commented on src/render/direct3d/SDL_render_d3d.c:
```diff
@@ -1552,13 +1552,32 @@ static int D3D_Reset(SDL_Renderer *renderer)
+    case 4:
+        PresentationInterval = D3DPRESENT_INTERVAL_FOUR;
+    default:
+        PresentationInterval = 0;
```
This works because the `!(caps.PresentationIntervals & PresentationInterval)` test below fails, but would it be clearer to just reutrn SDL_Unsupported() here?

slouken replied:
Yep!

sezero commented:
First crude attempt at a sdl2-compat patch: [patch.txt](https://github.com/libsdl-org/SDL/files/15299676/patch.txt)

<details>

```diff
diff --git a/src/sdl2_compat.c b/src/sdl2_compat.c
index 9e94502..689f55a 100644
--- a/src/sdl2_compat.c
+++ b/src/sdl2_compat.c
@@ -789,6 +789,7 @@ BOOL WINAPI _DllMainCRTStartup(HANDLE dllhandle, DWORD reason, LPVOID reserved)
 
 #define SDL2_RENDERER_SOFTWARE      0x00000001
 #define SDL2_RENDERER_ACCELERATED   0x00000002
+#define SDL2_RENDERER_PRESENTVSYNC  0x00000004
 #define SDL2_RENDERER_TARGETTEXTURE 0x00000008
 
 /* Events changed in SDL3; notably, the `timestamp` field moved from
@@ -3836,6 +3837,7 @@ DECLSPEC int SDLCALL
 SDL_GetRendererInfo(SDL_Renderer *renderer, SDL2_RendererInfo *info)
 {
     SDL_RendererInfo info3;
+    SDL_PropertiesID props;
     unsigned int i;
 
     SDL3_zerop(info);
@@ -3843,11 +3845,21 @@ SDL_GetRendererInfo(SDL_Renderer *renderer, SDL2_RendererInfo *info)
         return -1;
     }
 
+    props = SDL3_GetRendererProperties(renderer);
+
     info->name = info3.name;
-    info->flags = info3.flags;
     info->num_texture_formats = info3.num_texture_formats;
-    info->max_texture_width = info3.max_texture_width;
-    info->max_texture_height = info3.max_texture_height;
+    info->max_texture_width = (int)SDL3_GetNumberProperty(props, SDL_PROP_RENDERER_MAX_TEXTURE_SIZE_NUMBER, 0);
+    info->max_texture_height = info->max_texture_width;
+
+    if (SDL3_strcmp(info->name, "software") == 0) {
+        info->flags = SDL2_RENDERER_SOFTWARE | SDL2_RENDERER_TARGETTEXTURE;
+    } else {
+        info->flags = SDL2_RENDERER_ACCELERATED | SDL2_RENDERER_TARGETTEXTURE;
+    }
+    if (SDL3_GetNumberProperty(props, SDL_PROP_RENDERER_VSYNC_NUMBER, 0)) {
+        info->flags |= SDL2_RENDERER_PRESENTVSYNC;
+    }
 
     if (info->num_texture_formats > 16) {
         info->num_texture_formats = 16;
@@ -3872,22 +3884,22 @@ SDL_GetRenderDriverInfo(int idx, SDL2_RendererInfo *info)
 
     /* these are the values that SDL2 returns. */
     if ((SDL3_strcmp(name, "opengl") == 0) || (SDL3_strcmp(name, "opengles2") == 0)) {
-        info->flags = SDL2_RENDERER_ACCELERATED | SDL_RENDERER_PRESENTVSYNC | SDL2_RENDERER_TARGETTEXTURE;
+        info->flags = SDL2_RENDERER_ACCELERATED | SDL2_RENDERER_PRESENTVSYNC | SDL2_RENDERER_TARGETTEXTURE;
         info->num_texture_formats = 4;
         info->texture_formats[0] = SDL_PIXELFORMAT_ARGB8888;
         info->texture_formats[1] = SDL_PIXELFORMAT_ABGR8888;
         info->texture_formats[2] = SDL_PIXELFORMAT_XRGB8888;
         info->texture_formats[3] = SDL_PIXELFORMAT_XBGR8888;
     } else if (SDL3_strcmp(name, "opengles") == 0) {
-        info->flags = SDL2_RENDERER_ACCELERATED | SDL_RENDERER_PRESENTVSYNC;
+        info->flags = SDL2_RENDERER_ACCELERATED | SDL2_RENDERER_PRESENTVSYNC;
         info->num_texture_formats = 1;
         info->texture_formats[0] = SDL_PIXELFORMAT_ABGR8888;
     } else if (SDL3_strcmp(name, "direct3d") == 0) {
-        info->flags = SDL2_RENDERER_ACCELERATED | SDL_RENDERER_PRESENTVSYNC | SDL2_RENDERER_TARGETTEXTURE;
+        info->flags = SDL2_RENDERER_ACCELERATED | SDL2_RENDERER_PRESENTVSYNC | SDL2_RENDERER_TARGETTEXTURE;
         info->num_texture_formats = 1;
         info->texture_formats[0] = SDL_PIXELFORMAT_ARGB8888;
     } else if (SDL3_strcmp(name, "direct3d11") == 0) {
-        info->flags = SDL2_RENDERER_ACCELERATED | SDL_RENDERER_PRESENTVSYNC | SDL2_RENDERER_TARGETTEXTURE;
+        info->flags = SDL2_RENDERER_ACCELERATED | SDL2_RENDERER_PRESENTVSYNC | SDL2_RENDERER_TARGETTEXTURE;
         info->num_texture_formats = 6;
         info->texture_formats[0] = SDL_PIXELFORMAT_ARGB8888;
         info->texture_formats[1] = SDL_PIXELFORMAT_XRGB8888;
@@ -3896,7 +3908,7 @@ SDL_GetRenderDriverInfo(int idx, SDL2_RendererInfo *info)
         info->texture_formats[4] = SDL_PIXELFORMAT_NV12;
         info->texture_formats[5] = SDL_PIXELFORMAT_NV21;
     } else if (SDL3_strcmp(name, "direct3d12") == 0) {
-        info->flags = SDL2_RENDERER_ACCELERATED | SDL_RENDERER_PRESENTVSYNC | SDL2_RENDERER_TARGETTEXTURE;
+        info->flags = SDL2_RENDERER_ACCELERATED | SDL2_RENDERER_PRESENTVSYNC | SDL2_RENDERER_TARGETTEXTURE;
         info->num_texture_formats = 6;
         info->texture_formats[0] = SDL_PIXELFORMAT_ARGB8888;
         info->texture_formats[1] = SDL_PIXELFORMAT_XRGB8888;
@@ -3907,7 +3919,7 @@ SDL_GetRenderDriverInfo(int idx, SDL2_RendererInfo *info)
         info->max_texture_width = 16384;
         info->max_texture_height = 16384;
     } else if (SDL3_strcmp(name, "metal") == 0) {
-        info->flags = SDL2_RENDERER_ACCELERATED | SDL_RENDERER_PRESENTVSYNC | SDL2_RENDERER_TARGETTEXTURE;
+        info->flags = SDL2_RENDERER_ACCELERATED | SDL2_RENDERER_PRESENTVSYNC | SDL2_RENDERER_TARGETTEXTURE;
         info->num_texture_formats = 6;
         info->texture_formats[0] = SDL_PIXELFORMAT_ARGB8888;
         info->texture_formats[1] = SDL_PIXELFORMAT_ABGR8888;
@@ -3916,9 +3928,9 @@ SDL_GetRenderDriverInfo(int idx, SDL2_RendererInfo *info)
         info->texture_formats[4] = SDL_PIXELFORMAT_NV12;
         info->texture_formats[5] = SDL_PIXELFORMAT_NV21;
     } else if (SDL3_strcmp(name, "software") == 0) {
-        info->flags = SDL2_RENDERER_SOFTWARE | SDL_RENDERER_PRESENTVSYNC | SDL2_RENDERER_TARGETTEXTURE;
+        info->flags = SDL2_RENDERER_SOFTWARE | SDL2_RENDERER_PRESENTVSYNC | SDL2_RENDERER_TARGETTEXTURE;
     } else {  /* this seems reasonable if something currently-unknown shows up in SDL3. */
-        info->flags = SDL2_RENDERER_ACCELERATED | SDL_RENDERER_PRESENTVSYNC | SDL2_RENDERER_TARGETTEXTURE;
+        info->flags = SDL2_RENDERER_ACCELERATED | SDL2_RENDERER_PRESENTVSYNC | SDL2_RENDERER_TARGETTEXTURE;
         info->num_texture_formats = 1;
         info->texture_formats[0] = SDL_PIXELFORMAT_ARGB8888;
     }
@@ -3941,7 +3953,7 @@ DECLSPEC SDL_Renderer *SDLCALL
 SDL_CreateRenderer(SDL_Window *window, int idx, Uint32 flags)
 {
     SDL_PropertiesID props;
-    SDL_Renderer *renderer = NULL;
+    SDL_Renderer *renderer;
     const char *name = NULL;
 
     if (idx != -1) {
@@ -3966,12 +3978,14 @@ SDL_CreateRenderer(SDL_Window *window, int idx, Uint32 flags)
         name = SDL_SOFTWARE_RENDERER;
     }
 
-    flags &= ~(SDL2_RENDERER_SOFTWARE | SDL2_RENDERER_ACCELERATED | SDL2_RENDERER_TARGETTEXTURE); /* clear flags removed in SDL3 */
-    renderer = SDL3_CreateRenderer(window, name, flags);
+    renderer = SDL3_CreateRenderer(window, name);
     props = SDL3_GetRendererProperties(renderer);
     if (props) {
         SDL3_SetBooleanProperty(props, PROP_RENDERER_BATCHING, SDL2Compat_GetHintBoolean("SDL_RENDER_BATCHING", SDL_FALSE));
     }
+    if (flags & SDL2_RENDERER_PRESENTVSYNC) {
+        SDL3_SetRenderVSync(renderer, 1);
+    }
 
     return renderer;
 }
diff --git a/src/sdl3_syms.h b/src/sdl3_syms.h
index 48604b9..47be816 100644
--- a/src/sdl3_syms.h
+++ b/src/sdl3_syms.h
@@ -302,7 +302,7 @@ SDL3_SYM_RENAMED(SDL_bool,IntersectRectAndLine,GetRectAndLineIntersection,(const
 SDL3_SYM_PASSTHROUGH(int,GetNumRenderDrivers,(void),(),return)
 SDL3_SYM(const char *,GetRenderDriver,(int a),(a),return);
 SDL3_SYM(int,CreateWindowAndRenderer,(const char *a, int b, int c, SDL_WindowFlags d, SDL_Window **e, SDL_Renderer **f),(a,b,c,d,e,f),return)
-SDL3_SYM(SDL_Renderer*,CreateRenderer,(SDL_Window *a, const char *b, SDL_RendererFlags c),(a,b,c),return)
+SDL3_SYM(SDL_Renderer*,CreateRenderer,(SDL_Window *a, const char *b),(a,b),return)
 SDL3_SYM_PASSTHROUGH(SDL_Renderer*,CreateSoftwareRenderer,(SDL_Surface *a),(a),return)
 SDL3_SYM_PASSTHROUGH(SDL_Renderer*,GetRenderer,(SDL_Window *a),(a),return)
 SDL3_SYM(int,GetRendererInfo,(SDL_Renderer *a, SDL_RendererInfo *b),(a,b),return)
```

</details>

icculus commented on src/render/vulkan/SDL_render_vulkan.c:
```diff
@@ -4024,13 +4025,16 @@ static int VULKAN_SetVSync(SDL_Renderer *renderer, const int vsync)
+        /* Supported */
+        break;
+    default:
+        return SDL_Unsupported();
```
A vsync value of -1 in SDL would be VK_PRESENT_MODE_FIFO_RELAXED_KHR in Vulkan. I haven't parsed through all of `VULKAN_CreateSwapChain` but that would be where this would get added. If that's out of scope for this PR, we should open an issue so we remember to deal with it later.

slouken replied:
This is in, as a separate commit.

icculus commented:
> First crude attempt at a sdl2-compat patch: [patch.txt](https://github.com/libsdl-org/SDL/files/15299676/patch.txt)

Pedantically, the SDL_strcmp against "software" should be against SDL_SOFTWARE_RENDERER instead, but it comes to the same thing.

(I'm also not sure SDL should be in the business of publishing backend names in its public headers, but the software renderer is probably a special case.)

icculus commented:
My notes above aside, rest of the patches (for both SDL3 and sdl2-compat) look good to me. Surprised how much code this ended up touching, though!

sezero commented:
> > First crude attempt at a sdl2-compat patch: [patch.txt](https://github.com/libsdl-org/SDL/files/15299676/patch.txt)
> 
> Pedantically, the SDL_strcmp against "software" should be against SDL_SOFTWARE_RENDERER instead, but it comes to the same thing.

OK, changed my local copy

sezero commented:
> My notes above aside, rest of the patches (for [....] sdl2-compat) look good to me

Is calling SDL3_SetRenderVSync() with 1 as vsync value correct in there?

icculus commented:
> Is calling SDL3_SetRenderVSync() with 1 as vsync value correct in there?

Yes.

slouken commented:
> > Is calling SDL3_SetRenderVSync() with 1 as vsync value correct in there?
> 
> Yes.

Yup, your patch looks good.

icculus commented:
Removed SDL_RendererFlags and SDL_RENDERER_PRESENTVSYNC from the SDL3 wiki.
